<a href="https://colab.research.google.com/github/RinDataz/IBM-AI-Engineer-/blob/main/Project%20course%202.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install keras


In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam

# A. Build a baseline model (5 marks)

In [4]:
# Assuming `data` is a pandas DataFrame with the feature columns and the target column 'concrete_strength'
# Replace 'data' with your actual DataFrame

data = pd.read_csv('concrete_data.csv')
# Define the features and target
X = data.drop(columns=['Strength'])
y = data['Strength']

# Define a function to create and compile the model
def create_model():
    model = Sequential()
    model.add(Dense(10, input_dim=X.shape[1], activation='relu'))
    model.add(Dense(1))
    model.compile(optimizer=Adam(), loss='mean_squared_error')
    return model

# Store the mean squared errors
mse_list = []

# Repeat the process 50 times
for _ in range(50):
    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=None)

    # Create and train the model
    model = create_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)

    # Predict on the test set
    y_pred = model.predict(X_test)

    # Compute the mean squared error
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)

# Compute the mean and standard deviation of the mean squared errors
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"Mean MSE: {mean_mse}")
print(f"Standard Deviation of MSE: {std_mse}")


10/10 [==============================] - 0s 2ms/step
Mean MSE: 287.9223154152938
Standard Deviation of MSE: 277.8634778668784


**Mean MSE:** 287.9223154152938

**Standard Deviation of MSE:** 277.8634778668784


# B. Normalize the data (5 marks)

In [5]:
# Normalize the data
X_normalized = (X - X.mean()) / X.std()

# Define a function to create and compile the model with normalized data
def create_normalized_model():
    model = Sequential()
    model.add(Dense(10, input_dim=X_normalized.shape[1], activation='relu'))
    model.add(Dense(1))
    model.compile(optimizer=Adam(), loss='mean_squared_error')
    return model

# Store the mean squared errors for normalized data
normalized_mse_list = []

# Repeat the process 50 times for normalized data
for _ in range(50):
    # Split the normalized data into training and testing sets
    X_train_norm, X_test_norm, y_train_norm, y_test_norm = train_test_split(X_normalized, y, test_size=0.3, random_state=None)

    # Create and train the model with normalized data
    normalized_model = create_normalized_model()
    normalized_model.fit(X_train_norm, y_train_norm, epochs=50, verbose=0)

    # Predict on the test set with normalized data
    y_pred_norm = normalized_model.predict(X_test_norm)

    # Compute the mean squared error for normalized data
    mse_norm = mean_squared_error(y_test_norm, y_pred_norm)
    normalized_mse_list.append(mse_norm)

# Compute the mean and standard deviation of the mean squared errors for normalized data
mean_mse_normalized = np.mean(normalized_mse_list)
std_mse_normalized = np.std(normalized_mse_list)

print(f"Mean MSE for normalized data: {mean_mse_normalized}")
print(f"Standard Deviation of MSE for normalized data: {std_mse_normalized}")


10/10 [==============================] - 0s 2ms/step
Mean MSE for normalized data: 355.54742622846663
Standard Deviation of MSE for normalized data: 120.13256149155097


 Mean MSE for normalized data: 355.54742622846663

 Standard Deviation of MSE for normalized data:  120.13256149155097

# C. Increate the number of epochs (5 marks)

In [6]:
# Store the mean squared errors for increased epochs
increased_epochs_mse_list = []

# Repeat the process 50 times for increased epochs
for _ in range(50):
    # Split the normalized data into training and testing sets
    X_train_norm, X_test_norm, y_train_norm, y_test_norm = train_test_split(X_normalized, y, test_size=0.3, random_state=None)

    # Create and train the model with increased epochs
    increased_epochs_model = create_normalized_model()
    increased_epochs_model.fit(X_train_norm, y_train_norm, epochs=100, verbose=0)

    # Predict on the test set with increased epochs
    y_pred_increased_epochs = increased_epochs_model.predict(X_test_norm)

    # Compute the mean squared error for increased epochs
    mse_increased_epochs = mean_squared_error(y_test_norm, y_pred_increased_epochs)
    increased_epochs_mse_list.append(mse_increased_epochs)

# Compute the mean and standard deviation of the mean squared errors for increased epochs
mean_mse_increased_epochs = np.mean(increased_epochs_mse_list)
std_mse_increased_epochs = np.std(increased_epochs_mse_list)

print(f"Mean MSE for increased epochs: {mean_mse_increased_epochs}")
print(f"Standard Deviation of MSE for increased epochs: {std_mse_increased_epochs}")


10/10 [==============================] - 0s 2ms/step
Mean MSE for increased epochs: 165.37019767987388
Standard Deviation of MSE for increased epochs: 18.57821338950984


Mean MSE for increased epochs: 165.37019767987388

Standard Deviation of MSE for increased epochs: 18.57821338950984

# D. Increase the number of hidden layers (5 marks)

In [7]:
# Define a function to create and compile the model with three hidden layers
def create_model_three_hidden_layers():
    model = Sequential()
    model.add(Dense(10, input_dim=X_normalized.shape[1], activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    model.compile(optimizer=Adam(), loss='mean_squared_error')
    return model

# Store the mean squared errors for three hidden layers
three_hidden_layers_mse_list = []

# Repeat the process 50 times for three hidden layers
for _ in range(50):
    # Split the normalized data into training and testing sets
    X_train_three_hidden_layers, X_test_three_hidden_layers, y_train_three_hidden_layers, y_test_three_hidden_layers = train_test_split(X_normalized, y, test_size=0.3, random_state=None)

    # Create and train the model with three hidden layers
    three_hidden_layers_model = create_model_three_hidden_layers()
    three_hidden_layers_model.fit(X_train_three_hidden_layers, y_train_three_hidden_layers, epochs=50, verbose=0)

    # Predict on the test set with three hidden layers
    y_pred_three_hidden_layers = three_hidden_layers_model.predict(X_test_three_hidden_layers)

    # Compute the mean squared error for three hidden layers
    mse_three_hidden_layers = mean_squared_error(y_test_three_hidden_layers, y_pred_three_hidden_layers)
    three_hidden_layers_mse_list.append(mse_three_hidden_layers)

# Compute the mean and standard deviation of the mean squared errors for three hidden layers
mean_mse_three_hidden_layers = np.mean(three_hidden_layers_mse_list)
std_mse_three_hidden_layers = np.std(three_hidden_layers_mse_list)

print(f"Mean MSE for three hidden layers: {mean_mse_three_hidden_layers}")
print(f"Standard Deviation of MSE for three hidden layers: {std_mse_three_hidden_layers}")


10/10 [==============================] - 0s 2ms/step
Mean MSE for three hidden layers: 128.93370039880992
Standard Deviation of MSE for three hidden layers: 13.519949465499607


Mean MSE for three hidden layers: 128.93370039880992

Standard Deviation of MSE for three hidden layers: 13.519949465499607